File: backend/app/models/scene.py
from sqlalchemy import Column, Integer, String, JSON, Boolean, DateTime, Text, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.core.database import Base


class Scene(Base):
    """场景主表"""
    __tablename__ = "scenes"

    id = Column(Integer, primary_key=True, index=True)
    scene_code = Column(String(100), unique=True, index=True, nullable=False)
    scene_name = Column(String(200), nullable=False)
    description = Column(Text)
    keywords = Column(JSON, nullable=False, default=list)
    priority = Column(Integer, default=10)
    is_active = Column(Boolean, default=True)
    
    # 业务字段
    intent_type = Column(String(50))
    prompt_code = Column(String(100), index=True)  # 提示词编码
    action_type = Column(String(50))
    required_tools = Column(JSON, nullable=False, default=list)
    available_tools = Column(JSON, nullable=False, default=list)
    pre_action_steps = Column(JSON, nullable=False, default=list)
    post_action_steps = Column(JSON, nullable=False, default=list)
    
    # 树形结构字段
    type = Column(String(20), nullable=False, default='scene', index=True)  # center/business/scene
    parent_id = Column(Integer, ForeignKey('scenes.id'), nullable=True, index=True)
    
    # 通用配置字段 - 用于存储中心域和业务域的特定信息
    config = Column(JSON, nullable=False, default=dict)  # 灵活的配置字段
    
    # 公共字段（与其他模型保持一致）
    version = Column(Integer, default=1)
    created_by = Column(String(100))
    updated_by = Column(String(100))
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # 自关联关系
    parent = relationship('Scene', remote_side=[id], backref='children')
    
    # 关联版本历史
    history = relationship("SceneHistory", back_populates="scene", cascade="all, delete-orphan")

    def to_dict(self):
        return {
            "id": self.id,
            "sceneCode": self.scene_code,
            "sceneName": self.scene_name,
            "description": self.description,
            "keywords": self.keywords,
            "priority": self.priority,
            "isActive": self.is_active,
            "intentType": self.intent_type,
            "promptCode": self.prompt_code,
            "actionType": self.action_type,
            "requiredTools": self.required_tools,
            "availableTools": self.available_tools,
            "preActionSteps": self.pre_action_steps,
            "postActionSteps": self.post_action_steps,
            "type": self.type,
            "parentId": self.parent_id,
            "config": self.config,
            "version": self.version,
            "createdBy": self.created_by,
            "updatedBy": self.updated_by,
            "createdAt": self.created_at.isoformat() if self.created_at else None,
            "updatedAt": self.updated_at.isoformat() if self.updated_at else None
        }

    def to_tree_node(self):
        """转换为树节点格式"""
        return {
            "id": self.id,
            "sceneCode": self.scene_code,
            "sceneName": self.scene_name,
            "description": self.description,
            "keywords": self.keywords,
            "type": self.type,
            "priority": self.priority,
            "isActive": self.is_active,
            "promptCode": self.prompt_code,
            "config": self.config,
            "children": []
        }


class SceneHistory(Base):
    """场景版本历史表"""
    __tablename__ = "scene_history"

    id = Column(Integer, primary_key=True, index=True)
    scene_id = Column(Integer, ForeignKey("scenes.id"), nullable=False, index=True)
    scene_code = Column(String(100), nullable=False, index=True)  # 冗余字段，方便查询
    version = Column(Integer, nullable=False)
    
    # 场景数据快照
    scene_name = Column(String(200), nullable=False)
    description = Column(Text)
    keywords = Column(JSON, nullable=False, default=list)
    priority = Column(Integer, default=10)
    is_active = Column(Boolean, default=True)
    intent_type = Column(String(50))
    prompt_code = Column(String(100))  # 提示词编码快照
    action_type = Column(String(50))
    required_tools = Column(JSON, nullable=False, default=list)
    available_tools = Column(JSON, nullable=False, default=list)
    pre_action_steps = Column(JSON, nullable=False, default=list)
    post_action_steps = Column(JSON, nullable=False, default=list)
    type = Column(String(20), nullable=False, default='scene')
    parent_id = Column(Integer, nullable=True)
    config = Column(JSON, nullable=False, default=dict)  # 配置信息快照
    
    # 变更记录
    change_note = Column(Text)  # 变更说明
    created_by = Column(String(100))
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    scene = relationship("Scene", back_populates="history")

    def to_dict(self):
        return {
            "id": self.id,
            "sceneId": self.scene_id,
            "sceneCode": self.scene_code,
            "version": self.version,
            "sceneName": self.scene_name,
            "description": self.description,
            "keywords": self.keywords,
            "priority": self.priority,
            "isActive": self.is_active,
            "intentType": self.intent_type,
            "promptCode": self.prompt_code,
            "actionType": self.action_type,
            "requiredTools": self.required_tools,
            "availableTools": self.available_tools,
            "preActionSteps": self.pre_action_steps,
            "postActionSteps": self.post_action_steps,
            "type": self.type,
            "parentId": self.parent_id,
            "config": self.config,
            "changeNote": self.change_note,
            "createdBy": self.created_by,
            "createdAt": self.created_at.isoformat() if self.created_at else None
        }

File: backend-app/src/main/java/com/sitech/prodai/domain/entity/Scene.java
package com.sitech.prodai.domain.entity;

import com.sitech.prodai.common.JsonConverters;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

/**
 * 场景主表 —— 对齐 Python {@code app/models/scene.py::Scene}。
 *
 * <p>树形结构：center（中心） → business（业务） → scene（场景）。
 */
@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(name = "scenes", indexes = {
        @Index(name = "idx_scene_code", columnList = "scene_code", unique = true),
        @Index(name = "idx_scene_type", columnList = "type"),
        @Index(name = "idx_scene_parent_id", columnList = "parent_id"),
        @Index(name = "idx_scene_prompt_code", columnList = "prompt_code")
})
@EntityListeners(AuditingEntityListener.class)
public class Scene {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(name = "scene_code", unique = true, nullable = false, length = 100)
    private String sceneCode;

    @Column(name = "scene_name", nullable = false, length = 200)
    private String sceneName;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "keywords", nullable = false, columnDefinition = "TEXT")
    @Convert(converter = JsonConverters.JsonListConverter.class)
    private List<Object> keywords;

    @Column(name = "priority")
    private Integer priority = 10;

    @Column(name = "is_active")
    private Boolean isActive = true;

    @Column(name = "prompt_code", length = 100)
    private String promptCode;

    /** center / business / scene */
    @Column(name = "type", nullable = false, length = 20)
    private String type = "scene";

    @Column(name = "parent_id")
    private Integer parentId;

    @ManyToOne
    @JoinColumn(name = "parent_id", referencedColumnName = "id", insertable = false, updatable = false)
    private Scene parent;

    @OneToMany(mappedBy = "parent", cascade = CascadeType.ALL)
    private List<Scene> children;

    @OneToMany(mappedBy = "scene", cascade = CascadeType.ALL)
    private List<SceneHistory> history;

    /** 通用配置字段，存储中心域/业务域的特定信息 */
    @Column(name = "config", nullable = false, columnDefinition = "TEXT")
    @Convert(converter = JsonConverters.JsonMapConverter.class)
    private Map<String, Object> config;

    @Column(name = "version")
    private Integer version = 1;

    @Column(name = "created_by", length = 100)
    private String createdBy;

    @Column(name = "updated_by", length = 100)
    private String updatedBy;

    @CreatedDate
    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}
